core/src/main/java/org/mule/runtime/core/internal/value/cache/ValueProviderCacheId.java

```diff
@@ -0,0 +1,142 @@
+ * all the relevant parts involved in the resolution mechanism.
+ *
+ * A {@link ValueProviderCacheId} is expected to be immutable, except from it's attributes, which can be modified on any instance.
+ * They are intended to add additional information regarding how the {@link ValueProviderCacheId} was generated, but will not
+ * affect the value computation.
+ *
+ * @since 4.2.3, 4.3.0
+ */
+public class ValueProviderCacheId {
+
+  private final String sourceElementName;
+  private final String value;
+  private final List<ValueProviderCacheId> parts;
+  private final Map<String, String> attributes;
+
+  private ValueProviderCacheId(String sourceElementName, String value, List<ValueProviderCacheId> parts,
+                               Map<String, String> attributes) {
+    this.sourceElementName = sourceElementName;
+    this.parts = new ArrayList<>(parts);
+    this.attributes = attributes;
+    this.value = value + parts.stream().map(ValueProviderCacheId::getValue).reduce((a, b) -> a + b).orElse(EMPTY);
+  }
+
+  public List<ValueProviderCacheId> getParts() {
+    return this.parts;
+  }
+
+  public String getSourceElementName() {
+    return this.sourceElementName;
+  }
+
+  public String getValue() {
+    return this.value;
+  }
+
+  public Map<String, String> getAttributes() {
+    return this.attributes;
+  }
+
+  @Override
+  public String toString() {
+    StringBuilder stringBuilder = new StringBuilder();
+    stringBuilder
+        .append(sourceElementName)
+        .append(" : ")
+        .append(value)
+        .append(", attributes: {");
+    this.attributes.forEach((k, v) -> stringBuilder.append(lineSeparator()).append(k).append(" : ").append(value));
+    stringBuilder.append("}");
+    return stringBuilder.toString();
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hashCode(value);
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (obj instanceof ValueProviderCacheId) {
+      return Objects.equals(value, ((ValueProviderCacheId) obj).value);
+    }
+    return false;
+  }
+
+  public static class ValueProviderCacheIdBuilder {
+
+    private String sourceElementName;
+    private String hashValue;
+    private List<ValueProviderCacheId> parts;
+    private Map<String, String> attributes;
+
+    public static ValueProviderCacheId aValueProviderCacheId(ValueProviderCacheIdBuilder builder) {
+      return builder.build();
+    }
+
+    private ValueProviderCacheIdBuilder() {
+      this.parts = new LinkedList<>();
+      this.attributes = new HashMap<>();
+    }
+
+    public static ValueProviderCacheIdBuilder fromElementWithName(String sourceElementName) {
+      ValueProviderCacheIdBuilder builder = new ValueProviderCacheIdBuilder();
+      builder.sourceElementName = sourceElementName;
+      return builder;
+    }
+
+    public ValueProviderCacheIdBuilder withHashValueFrom(Object hashedObject) {
```

Comment:
The fact that we are using the string representation for every hashedValue is a coincidence. It may not be the same in the future.  Besides, Objects.hash() already works with Object instances, so there is no need to downcast or anything weird.